### PR TITLE
Move from Ivory to FOSCkeditor

### PR DESCRIPTION
**Pull request by kunicmarko20**
I am targeting this branch, because this is a bc fix that would allow formatter with sf4

## Changelog

```markdown
### Deprecated
- using of IvoryCKEditorBundle
```

## To do

- [x] Wait for stable release of [FOSCKEditorBundle](https://github.com/FriendsOfSymfony/FOSCKEditorBundle)
- [x] Update tests

## Subject

We took over IvoryCKEditorBundle and now it is under FOS, in the next few days there will be a stable release, this is just to check if tests pass.

### CODE REVIEW

**Comment by kunicmarko20** [comment]
DO NOT MERGE.

File: src/Form/Type/SimpleFormatterType.php [hunk 1]
```diff
@@ -44,10 +44,10 @@ class SimpleFormatterType extends AbstractType
...
 
     /**
-     * @param ConfigManagerInterface         $configManager    An Ivory CKEditor bundle configuration manager
-     * @param PluginManagerInterface|null    $pluginManager    An Ivory CKEditor bundle plugin manager
-     * @param TemplateManagerInterface|null  $templateManager  An Ivory CKEditor bundle template manager
-     * @param StylesSetManagerInterface|null $stylesSetManager An Ivory CKEditor bundle styles set manager
+     * @param ConfigManagerInterface         $configManager    An FOS CKEditor bundle configuration manager
+     * @param PluginManagerInterface|null    $pluginManager    An FOS CKEditor bundle plugin manager
+     * @param TemplateManagerInterface|null  $templateManager  An FOS CKEditor bundle template manager
+     * @param StylesSetManagerInterface|null $stylesSetManager An FOS CKEditor bundle styles set manager
```
Comment (greg0ire) [inline src/Form/Type/SimpleFormatterType.php]: `s/An/A`

Comment (greg0ire) [reply src/Form/Type/SimpleFormatterType.php]: Same above

**Comment by kunicmarko20** [comment]
@core23 Well, we provided an abstraction to Ivory CKEditor, so if anyone was depending on it without explicitly adding it as a requirement, it is not our fault? I discussed this with @greg0ire

**Comment by kunicmarko20** [comment]
And we are green ✅, cc @sonata-project/contributors

**Comment by kunicmarko20** [comment]
hm, you mean if you are extending some of formatter classes and you change something? Ye, I think that would be a BC break. Should we move this to master then and do a major after we clean the master branch?

**Comment by greg0ire** [comment]
Wait what? Which classes? Can you give an example please?

**Comment by kunicmarko20** [comment]
If someone extended some of our services and overwrote it as a services, he would have to typehint agents IvoryCKEditor namespace, their app would brake with this change as our classes are not final and they can do that.

**Comment by greg0ire** [comment]
I see. How about using `class_alias` in Sonata, or maybe even in your bundle?

**Comment by kunicmarko20** [comment]
We decided we didn't want to do that.

So, I rebase on master and we consider doing a major of formatter?

**Comment by kunicmarko20** [comment]
And not sure class alias for 3rd party should be in sonata?

**Comment by greg0ire** [comment]
Yeah you're right, we probably shouldn't define aliases for our other classes… so let's go with a major I guess :'(

**Comment by kunicmarko20** [comment]
Will rebase to master, clean master and make everything final? Does that sound ok?

File: src/Form/Type/FormatterType.php [hunk 2]
```diff
@@ -57,9 +57,9 @@ class FormatterType extends AbstractType
     private $templateManager;
 
     /**
-     * @param ConfigManagerInterface|null   $configManager   An Ivory CKEditor bundle configuration manager
-     * @param PluginManagerInterface|null   $pluginManager   An Ivory CKEditor bundle plugin manager
-     * @param TemplateManagerInterface|null $templateManager An Ivory CKEditor bundle template manager
+     * @param ConfigManagerInterface|null   $configManager   A FOS CKEditor bundle configuration manager
+     * @param PluginManagerInterface|null   $pluginManager   A FOS CKEditor bundle plugin manager
+     * @param TemplateManagerInterface|null $templateManager A FOS CKEditor bundle template manager
```
Comment (greg0ire) [inline src/Form/Type/FormatterType.php]: Please allow FOS or Ivory here

File: src/DependencyInjection/SonataFormatterExtension.php [hunk 3]
```diff
@@ -41,7 +41,7 @@ public function load(array $configs, ContainerBuilder $container)
 
         $bundles = $container->getParameter('kernel.bundles');
 
-        if (isset($bundles['IvoryCKEditorBundle'])) {
+        if (isset($bundles['FOSCKEditorBundle'])) {
```
Comment (greg0ire) [inline src/DependencyInjection/SonataFormatterExtension.php]: Please check both bundles, and maybe throw if both are registered.

File: composer.json [hunk 4]
```diff
@@ -24,7 +24,7 @@
     ],
     "require": {
         "php": "^5.6 || ^7.0",
-        "egeloen/ckeditor-bundle": "^4.0 || ^5.0 || ^6.0",
```
Comment (greg0ire) [inline composer.json]: Please restore this

File: composer.json [hunk 5]
```diff
@@ -24,7 +24,7 @@
...
+        "friendsofsymfony/ckeditor-bundle": "^1.0",
```
Comment (greg0ire) [inline composer.json]: Maybe move this to suggest?

File: src/DependencyInjection/SonataFormatterExtension.php (hunk 3 above)
Comment (greg0ire) [reply src/DependencyInjection/SonataFormatterExtension.php]: Please trigger a deprecation iff Ivory is registered

File: src/Resources/config/form.xml [hunk 6]
```diff
@@ -5,16 +5,16 @@
             <tag name="form.type" alias="sonata_formatter_type"/>
             <argument type="service" id="sonata.formatter.pool"/>
             <argument type="service" id="translator"/>
-            <argument type="service" id="ivory_ck_editor.config_manager"/>
-            <argument type="service" id="ivory_ck_editor.plugin_manager"/>
-            <argument type="service" id="ivory_ck_editor.template_manager"/>
+            <argument type="service" id="fos_ck_editor.config_manager"/>
+            <argument type="service" id="fos_ck_editor.plugin_manager"/>
+            <argument type="service" id="fos_ck_editor.template_manager"/>
```
Comment (greg0ire) [inline src/Resources/config/form.xml]: Please make this dynamic somehow (either do it in php or load an xml file conditionally)

File: src/SonataFormatterBundle.php [hunk 7]
```diff
@@ -30,7 +30,7 @@ public function boot()
     public function registerFormMapping()
     {
         FormHelper::registerFormTypeMapping([
-            'ckeditor' => 'Ivory\CKEditorBundle\Form\Type\CKEditorType',
+            'ckeditor' => 'FOS\CKEditorBundle\Form\Type\CKEditorType',
```
Comment (greg0ire) [inline src/SonataFormatterBundle.php]: Please make this dynamic

**Comment by greg0ire** [comment]
> Cant we move everything to require-dev and let developer choose between ivory or fos? In a BC way?

This will crash on update I'm afraid.

File: src/Form/Type/FormatterType.php (hunk 2 above)
Comment (kunicmarko20) [reply src/Form/Type/FormatterType.php]: Remove typehint and add an `if` in method?

File: src/DependencyInjection/SonataFormatterExtension.php (hunk 3 above)
Comment (kunicmarko20) [reply src/DependencyInjection/SonataFormatterExtension.php]: So if they install fos on their own, we should use fos config else we continue doing it like now but throw deprecation

Comment (kunicmarko20) [reply src/DependencyInjection/SonataFormatterExtension.php]: and on next major we will replace ivory with fos

File: src/Resources/config/form.xml (hunk 6 above)
Comment (kunicmarko20) [reply src/Resources/config/form.xml]: Make 2 `form.xml` one for fos, one for ivory?

File: composer.json (hunk 5 above)
Comment (kunicmarko20) [reply composer.json]: If I am going to check types, shouldn't it be in require dev?

**Comment by kunicmarko20** [comment]
Need to update tests and should I revert docs to ivory or should I mention both?

**Comment by kunicmarko20** [comment]
wouldn't this still force IvoryCKEditor because it is still in require part of composer? and it still fails to install on sf4 :disappointed:

**Comment by greg0ire** [comment]
> Need to update tests and should I revert docs to ivory or should I mention both?

Only mention the right way please.

**Comment by greg0ire** [comment]
> wouldn't this still force IvoryCKEditor because it is still in require part of composer? and it still fails to install on sf4

Ah that's right, I forgot that… but there is still hope: How about using the `replaces` key in FOS's composer.json?

**Comment by kunicmarko20** [comment]
TIL 😄 cc @ElectricMaxxx we should add `replace` in composer.json

**Comment by greg0ire** [comment]
Make sure to discuss it and not rush though, I'm not sure of the implications, since you changed the namespaces, it's not the same functionality, technically… but anyone using your package is aware they need to change the namespaces I suppose.

**Comment by ElectricMaxxx** [comment]
I currently do not understand the issue. As you do not expose a third party Interface without wrapping it, the user would not have any change, or am i right. When somebody relies on an Ivory class requiring your bundle only, the it is no BC from my POV.

**Comment by kunicmarko20** [comment]
@ElectricMaxxx we have typehints in constructors of our form types, if someone is extending and overriding that form types, his app would break if we change the type of class in extended class. So, I will check out that replace key later more or we can do a Major release maybe.

File: src/DependencyInjection/SonataFormatterExtension.php [hunk 8]
```diff
@@ -41,8 +41,10 @@ public function load(array $configs, ContainerBuilder $container)
 
         $bundles = $container->getParameter('kernel.bundles');
 
-        if (isset($bundles['IvoryCKEditorBundle'])) {
-            $loader->load('form.xml');
+        if (isset($bundles['FOSCKEditorBundle'])) {
+            $loader->load('form_fos.xml');
+        } elseif (isset($bundles['IvoryCKEditorBundle'])) {
+            $loader->load('form_ivory.xml');
```
[the remark this answers is not in the document]
Comment (kunicmarko20) [reply src/DependencyInjection/SonataFormatterExtension.php]: then I just remove stuff from ctor and throw 1 deprecation here?

[the remark this answers is not in the document]
Comment (greg0ire) [reply src/DependencyInjection/SonataFormatterExtension.php]: yup. People don't need to be told to move to FOS 23409823 times :P

**Comment by greg0ire** [comment]
> What about removing the type hint an do a manual instanceof, doing so you could easily deprecate that Argument type. Removing a type would be no BC break.

Uuuuh… that's exactly what I suggested and has been done

**Comment by kunicmarko20** [comment]
Seems that we are going to revert replace https://github.com/FriendsOfSymfony/FOSCKEditorBundle/pull/97 after all, so, should I rebase this on master? cc @sonata-project/contributors

**Comment by kunicmarko20** [comment]
Closing this as there is no way to implement this without BC Break, will just change it on master.

**Comment by greg0ire** [comment]
Whaaaat? Why? If you don't use `replace` , you can still get this, even if it means having both bundles install. Don't assume next major will happen soon, I've made that kind of assumptions in the pass, it's usually wrong.

File: docs/reference/installation.rst [hunk 9]
```diff
@@ -4,7 +4,20 @@
...
+
+.. code-block:: json
+
+    "replace": {
+        "egeloen/ckeditor-bundle": "*"
+    }
+
+And after that you need to install `FOSCKEditorBundle`_.
+
+Install SonataFormatterBundle:
```
Comment (kunicmarko20) [inline docs/reference/installation.rst]: This part is the only new thing.

**Comment by kunicmarko20** [comment]
Thanks to @greg0ire we found a way to keep this on stable and avoid BC Break.

**Comment by greg0ire** [comment]
The docs job failed

**Comment by kunicmarko20** [comment]
should be ok now

**Comment by greg0ire** [comment]
@core23 please unblock this

**Comment by greg0ire** [comment]
Thanks @kunicmarko20 !

**Comment by kunicmarko20** [comment]
It would be a bc break 😞

**Pull request merged**